### ISSUE TITLE
Member list not sorted logically

### ISSUE THREAD

**Issue by laszlovl** [conversejs/converse.js#1604]
In case of a MUC with permanent members, I would probably expect all online members to be listed first; or perhaps for members to be listed alphabetically regardless of their state. But whatever happens right now seems pretty random, check the attached screenshot.

I tried to look in the code to see what's going on exactly, but I struggled to find the code responsible for the exact sorting mechanism.

![memberlist](https://user-images.githubusercontent.com/12112486/59569165-f9f2cd80-9085-11e9-88b0-1ee1f9a94235.jpg)

**Comment by jcbrand** [comment]
Moderators are shown first (alphabetically), then the rest are shown (alphabetically).

**Comment by laszlovl** [comment]
I presume "moderator" is just an alias for "either owner or admin"? Not sure about the value of showing two separate tags for a single user either way. But in my screenshot, the bottom user has exactly the same rights as the 2nd user, so I don't know why they aren't considered a "moderator" and why he's shown all the way at the bottom.

Regardless, I feel like it's more useful to show all online users first; there's a huge difference between "online and receiving these messages right now" and "could theoretically join this room at some point". I might even list these offline participants under a completely separate header. Thoughts?

**Comment by jcbrand** [comment]
> I presume "moderator" is just an alias for "either owner or admin"?

Moderator is a role and owner/admin is an affiliation.

Roles are session-based. Affiliations are permanent (until changed).
All admins/owners are automatically moderators, but not vice versa.

Given my last sentence, I don't understand why the last occupant is not shown as a moderator as well. Perhaps this is a bug in Converse, I'm not sure.

> I might even list these offline participants under a completely separate header

I don't see how this necessarily reduces confusion.

**Comment by laszlovl** [comment]
Right. In that case it makes some sense, since the last user is not online I guess he's not granted the moderator role.

I think there's some room for improvement here, but I'll get a second opinion from a designer before getting back to you. Could you still point me at where in the code this sorting is determined?

**Comment by jcbrand** [comment]
Sorting happens here: https://github.com/conversejs/converse.js/blob/de3099a9d86c338ca17b29a4c70abc49df357452/src/headless/converse-muc.js#L1763

> Right. In that case it makes some sense, since the last user is not online I guess he's not granted the moderator role.

Yep, makes sense now.

I think we should just extend that comparator method to also take admin/owner affiliation into account, not just role.

**Comment by BuilovAlexey** [comment]
@laszlovl Hello :) You wrote "In case of a MUC with permanent members". How did you ensure that offline members of the group are not “removed” from the list of group members?
I work with prosody server. After the participants leave, they are displayed as offline, but if you restart yourself in the group, all "offline" are removed from the list and remain only online. I want them to stay on the list

**Comment by laszlovl** [comment]
@BuilovAlexey I'm not sure - I use Ejabberd and this just works out of the box. Adding someone as "member" to a MUC will keep that affiliation forever, also after they go offline or even a server restart.

**Comment by BuilovAlexey** [comment]
I use the Prosody server (may be bag inside him). Well, thanks a lot for the answer)

**Comment by jcbrand** [comment]
@laszlovl, @BuilovAlexey: Manually adding someone as a member, admin or owner makes them a permanent member of the MUC in Prosody as well.

Additionally in Prosody you can set `allow_unaffiliated_register = true` which will allow users to register themselves as members. In Converse.js you can then set [auto_register_muc_nickname](https://conversejs.org/docs/html/configuration.html#auto-register-muc-nickname) to true and all users who enter a MUC will automatically become permanent members.
